scaler/io/ymq/pymod_ymq/waiter.h

```
@@ -0,0 +1,116 @@
+            return *this;
+
+        this->_waiter   = other._waiter;
+        this->_ymqState = other._ymqState;
+        return *this;
+    }
+
+    Waiter& operator=(Waiter&& other) noexcept
+    {
+        if (this == &other)
+            return *this;
+
+        this->_waiter   = std::move(other._waiter);
+        this->_ymqState = other._ymqState;
+        other._ymqState = nullptr;  // invalidate the moved-from object
+        return *this;
+    }
+
+    void signal()
+    {
+        if (eventfd_write(*_waiter, 1) < 0) {
+            throw std::runtime_error("Failed to signal waiter: " + std::string(std::strerror(errno)));
+        }
+    }
+
+    // Block until signaled through `signal()` and return `true`.
+    //
+    // If interrupted by an OS signal and an exception is raised by a Python signal handler, returns `false`.
+    bool wait()
+    {
+        pollfd pfds[2] = {
+            {
+                .fd      = *_waiter,
+                .events  = POLLIN,
+                .revents = 0,
+            },
+            {
+                .fd      = _ymqState->signalWakeupFDWr,
+                .events  = POLLIN,
+                .revents = 0,
+            }};
+
+        for (;;) {
+            int ready = poll(pfds, 2, -1);
+            if (ready < 0) {
+                if (errno == EINTR)
+                    continue;
+                throw std::runtime_error("poll failed");
+            }
+
+            if (pfds[0].revents & POLLIN)
+                return true;  // signal() called
+
+            if (pfds[1].revents & POLLIN) {
+                // Interrupted by an OS signal. Run the Python signal handlers and wait again.
+                PyEval_RestoreThread(_ymqState->threadState);
+                if (PyErr_CheckSignals() == -1) {
+                    PyErr_SetString(
+                        *_ymqState->PyInterruptedExceptionType,
+                        "A synchronous YMQ operation was interrupted by a signal handler exception");
```

In this case I'm actually not sure if we need to raise an error here. My understanding is that the Python signal handler has raised an error, and therefore I think all we need to do if return. From the docs:

> if a Python signal handler raises an exception, the error indicator is set and the function returns -1 immediately

I think a good example of this is Ctrl+C -> KeyboardInterrupt